app/forms/idv/api_image_upload_form.rb

```diff
@@ -25,10 +26,24 @@ def submit
       FormResponse.new(
         success: valid?,
         errors: errors.messages,
-        extra: {},
+        extra: {
+          status: status,
+          remaining_attempts: remaining_attempts,
+        },
       )
     end
 
+    def status
+      return :ok if valid?
+      return :too_many_requests if errors.key?(:limit)
+      :bad_request
+    end
+
+    def remaining_attempts
+      return if valid?
+      Throttler::RemainingCount.call(document_capture_session.user_id, :idv_acuant)
```

Comment:
so I recently discovered that calling `valid?` resets all checks and calls everything again, so I think we may want to memoize a value in `throttle_if_rate_limited` -- otherwise we'll be double-counting each attempt

Reply:
> so I recently discovered that calling `valid?` resets all checks and calls everything again, so I think we may want to memoize a value in `throttle_if_rate_limited` -- otherwise we'll be double-counting each attempt

Good catch. I guess it also has me second-guessing whether a validation should be pure, since the increment behavior is most definitely a side-effect. Memoizing the throttle value seems to work well enough. I've changed to this in 88ddd43.

Alternatively, maybe it could make sense to separate the increment from the validation. If the idea is increment should be tied to a submission, maybe `throttled_else_increment` should be called directly from `submit`, and `throttle_if_rate_limited` consider the instance value assigned?

```rb
def submit
  throttled_else_increment

  # ...
end

# ...

def throttle_if_rate_limited
  return unless @throttled
  errors.add(:limit, t('errors.doc_auth.acuant_throttle'))
end

def throttled_else_increment
  @throttled = Throttler::IsThrottledElseIncrement.call(
    document_capture_session.user_id,
    :idv_acuant,
  )
end
```

Reply:
I'm a fan the approach above. Incrementing on submission instead of as a side-effect of validation makes more sense to me.

Reply:
Updated in da7a51f14.

It also has the added advantage to avoid a dependency between the validations, where previously incrementing via `throttle_if_rate_limited` would require a valid `document_capture_session`, but at the time of validation wasn't guaranteed to be assigned.

Reply:
Yeah I ran into the same issue in the RISC security events form: https://github.com/18F/identity-idp/blob/master/app/forms/security_event_form.rb#L121-L130

Maybe this is a case where we could/should break from ActiveModel validations to have something that only validates once? Other forms we've done do `@success = valid?` and then check the ivar